### PR TITLE
feat: js built in functions

### LINKED ISSUE #3150: Port more javascript builtin functions
### Describe the feature

Add bindings for missing builtin functions:

https://github.com/paradigmxyz/reth/blob/cf10590e4a8f2a35a1826149a746bedb8fd2e000/crates/revm/revm-inspectors/src/tracing/js/builtins.rs#L19-L19

reference impl:

https://github.com/ethereum/go-ethereum/blob/73697529994e14996b7740730481e926d5ec3e40/eth/tracers/js/goja.go#L372-L372

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by 0xJepsen**
Closes #3150 

Question for the isPrecompiles. I suspect the best way to do this would be to store the precompiles in the context `ctx` looking something like `let active_precompiles = get_active_precompiles(ctx); ` Do we have a reth primitive for them i could use?

### CODE REVIEW

**Review by mattsse — CHANGES_REQUESTED** [review]
sweet!

these look super clean.

you're right, I think we need to register the precompile function inside the `initialize_interp` hook we get access to the precompiles via `EVMData`

**Comment by 0xJepsen** [comment]
I can see where we get the precompiles to from `EVMData`. Are you suggesting we writing the `isPrecompiled` function *inside* the `initialize_interp` function? How do we add it to the context if we do it this way? 

One thing i was thinking is that I add the precompiles to the context struct and set that in the constructor. Then keep things where there are for the most part

Does that sound okay? Or did you have something else in mind?

**Comment by mattsse** [comment]
> I can see where we get the precompiles to from EVMData. Are you suggesting we writing the isPrecompiled function inside the initialize_interp function? 

yes

> How do we add it to the context if we do it this way?

we have the context as field

**Comment by 0xJepsen** [comment]
This is my attempt at this, i used a closure to capture the EVM state variable. It feels a little messy to me, maybe you will have some insight on how to clean it better.

File: crates/revm/revm-inspectors/src/tracing/js/builtins.rs [hunk 1]
```diff
@@ -122,6 +129,92 @@ pub(crate) fn to_bigint(value: U256, ctx: &mut Context<'_>) -> JsResult<JsValue>
...
+                    return Err(JsError::from_native(
+                        JsNativeError::typ().with_message("invalid salt hex string"),
+                    ))
+                }
+            }
+        }
+        Err(_) => {
+            return Err(JsError::from_native(JsNativeError::typ().with_message("invalid salt type")))
+        }
+    };
```
Comment (mattsse) [inline crates/revm/revm-inspectors/src/tracing/js/builtins.rs:158]: I think we already have this:

https://github.com/paradigmxyz/reth/blob/4d3d4974a48b4e838ccda50c3171da11c4885a13/crates/revm/revm-inspectors/src/tracing/js/builtins.rs#L133-L147

I believe this is equivalent

File: crates/revm/revm-inspectors/src/tracing/js/mod.rs [hunk 2]
```diff
@@ -277,9 +277,28 @@ where
...
+                let addr = bytes_to_address(buf);
+                for p in precompiles.keys() {
+                    if *p == addr.0 {
+                        return Ok(JsValue::from(true))
+                    }
+                }
+                Ok(JsValue::from(false))
+            };
+        self.ctx
+            .register_global_callable("isPrecompiled", 3, unsafe {
```
Comment (mattsse) [inline crates/revm/revm-inspectors/src/tracing/js/mod.rs:298]: I'd like to have a function for this in builtins.rs that takes the precompiles and installs the js function

Comment (mattsse) [inline crates/revm/revm-inspectors/src/tracing/js/mod.rs:298]:
```suggestion
            .register_global_builtin_callable("isPrecompiled", 1, unsafe {
```

**Review by mattsse — CHANGES_REQUESTED** [review]
nice, a few suggestions and requests, but this is very close

**Comment by 0xJepsen** [comment]
@mattsse I have been wrestling with this for a while and haven't made much headway. The problem i am having is that if we want to make the precompiles accessible via the `Context` we have to first turn them into a `JsValue` object which and then from the `builtins.rs` we then have the turn that `JsValue` back into an `iterator` or some value we can handle in rust, most of this involves making a variety of helper functions and feels really messy. Is there an alternative way we could make the precompiles available to the builtins other than through the js context? Or is that kinda our only option?

**Comment by mattsse** [comment]
we need something like we do here (although I don't think the Gc wrapper here is necessary)

https://github.com/paradigmxyz/reth/blob/0d691fee1697d18155cd23ccde0f1949ee90a289/crates/revm/revm-inspectors/src/tracing/js/bindings.rs#L553-L568

I think what we need for precompile is something like:

```rust
struct PrecompileWrapper(Precompiles)

impl PrecompileWrapper {

  pub fn register_callable(self, ctx: &mut Context) {
     let mut obj = ...;

     let ir_precompile = FunctionObjectBuilder::new(
            ctx,
            NativeFunction::from_copy_closure_with_captures(
                move |_this, args, precompiles, ctx| {
                    let val = args.get_or_undefined(0).clone();
                    // convert val to address 
                    // check if `precompiles` contains address
                    ....
                },
                self,
            ),
        )
        .length(1)
        .build();
  

     // register callable
  }   

}

unsafe impl Tracer for PrecompileWrapper {emptry_trace!()}

```

**Comment by mattsse** [comment]
@0xJepsen do you have bandwith to get this over the line?

If not, I'm happy to make the changes requested, there's not much missing.

**Comment by 0xJepsen** [comment]
> @0xJepsen do you have bandwith to get this over the line?
> 
> If not, I'm happy to make the changes requested, there's not much missing.

Will be hard for me this week as we just kicked off a big sprint on [Arbiter](https://github.com/primitivefinance/arbiter). Feel free to pick up where i left off, i'm actually very curious on you're approach. I appreciate the openness in Reth i've learned a lot already.

**Review by mattsse — APPROVED** [review]

**Pull request merged**